_posts/2024-03-28-fuse-storage-driver.adoc

```diff
@@ -0,0 +1,84 @@
+categories: []
+keywords: ['fuse', 'storage', 'driver']
+slug: /@david.kwon/fuse-storage-driver
+---
+
+Enable faster builds and storage savings with the fuse-overlayfs storage driver in Eclipse Che.
+
+Container images consist of https://docs.docker.com/build/guide/layers[layers] which are stored and used for building and running containers. A huge benefit of this layer anatomy is that, assuming that each image layer stores only the file differences compared to the previous layer (ie. the delta), each layer is small which generally allows time and space savings when building and running containers. This is because layers allow efficient caching and sharing between containers.
+
+A https://docs.docker.com/storage/storagedriver/select-storage-driver[storage driver] for Docker and Podman is what manages these image layers upon image pulling, building, and running. By default, Eclipse Che uses the vfs storage driver for Podman in the https://github.com/devfile/developer-images[Universal Development Image]. While vfs is generally considered very stable, the lack of https://en.wikipedia.org/wiki/Copy-on-write[copy-on-write] (CoW) support poses a significant disadvantage compared to other storage drivers like fuse-overlayfs, overlay2, and btrfs.
+
+## What's the disadvantage with vfs?
+
+As mentioned previously, having smaller layers allows time and space savings when building and running containers. The disadvantage with vfs's lack of CoW is that whenever a new layer is created, a complete copy of the previous layer is created. This can create duplicate and redundant data in each layer, and can quickly fill up your storage space especially when working with larger images. Comparing that with image layers created with CoW-supported storage drivers like fuse-overlayfs, those image layers would typically not contain redundant data and would remain smaller since only the delta is stored in each image layer.
+
+image::/assets/img/fuse-storage-driver/layer-diagram.png[]
+Fig. 1. Example comparison of image layer sizes between vfs and overlay storage drivers. The total size of the image layers created with overlay is about half the size of the image layers created with vfs.
+
+## How can we use fuse-overlayfs in an Eclipse Che cloud development environment?
+
+To use fuse-overlayfs in an Eclipse Che cloud development environment (CDE), the CDE's pod requires access to the `/dev/fuse` device from the host operating system. For Kubernetes, support for this can vary depending on the platform. However, for Openshift, version 4.15 allows access to the `/dev/fuse` device https://docs.openshift.com/container-platform/4.15/release_notes/ocp-4-15-release-notes.html#ocp-4-15-nodes-dev-fuse[without any cluster modifications]. Here's the documentation on how to https://eclipse.dev/che/docs/stable/end-user-guide/accessing-fuse[enable the overlay storage driver for CDEs running on OpenShift].
```

Comment:
> However, for Openshift, version 4.15 allows access to the /dev/fuse device [without any cluster modifications](https://docs.openshift.com/container-platform/4.15/release_notes/ocp-4-15-release-notes.html#ocp-4-15-nodes-dev-fuse).

IMHO, it gives an impression that there is something wrong with OCP older than 4.15, while, [according to the documentation](https://eclipse.dev/che/docs/stable/end-user-guide/accessing-fuse/), it's possible to make /dev/fuse accessible to workspace containers on OpenShift versions older than 4.15 as well.

Reply:
Fixed